### ISSUE TITLE
See what can be borrowed from similar projects

### ISSUE THREAD

**Issue by reubano** [nerevu/riko#10]
2 projects to look into initially are [pypes](https://github.com/fullscale/pypes) and [mario](https://github.com/colinmarc/mario).

**Comment by reubano** [comment]
- [x] [RssPercolator](https://github.com/olviko/RssPercolator) [[HN Comment](https://news.ycombinator.com/item?id=12137618)]
- [x] [plagger](https://github.com/miyagawa/plagger) [[HN Comment](https://news.ycombinator.com/item?id=12136409)]
- [x] [turtle](https://github.com/iopipe/turtle) [[HN Comment](https://news.ycombinator.com/item?id=12139946)]
- [x] [node-machine](https://github.com/node-machine/machine)

**Comment by reubano** [comment]
# Riko Roadmap

Status: Proposed
Research date: 2026-08-05
Primary research issue: [#10 - What can be borrowed from similar projects?](https://github.com/nerevu/riko/issues/10)
Target baseline: `features` branch, Riko 0.72.x

This roadmap converts the research backlog in issue #10 into an implementation sequence. It is a dependency-ordered plan, not a calendar commitment.

## Product direction

Riko should remain a small Python stream-processing library first. It should become easier to extend, validate, inspect, operate, and eventually visualize without turning the core package into a hosted workflow platform.

The intended progression is:

1. Make local pipelines correct and predictable.
2. Make modules self-describing.
3. Make third-party modules safe to discover and package.
4. Make workflow files versioned and fully validatable.
5. Add observability and operational policies.
6. Add adapters and optional execution drivers.
7. Build visual tooling as a separate consumer of the same contracts.

## Research conclusions

Issue #10 named six projects. Their implementations are old, but several design ideas remain useful.

| Project | Useful ideas | What Riko should not copy |
| --- | --- | --- |
| [Pypes](https://github.com/fullscale/pypes) | Flow-based applications as graphs of black-box components connected through named ports; explicit component interfaces; editor and runtime sharing one graph model. | Stackless Python, Python 2 assumptions, or a monolithic drag-and-drop application coupled to the runtime. |
| [Mario](https://github.com/colinmarc/mario) | Adapters for files, sockets, generators, subprocesses, and transform functions; explicit start/close behavior; handling partial chunks at boundaries. | A byte-pump abstraction as Riko's primary data model. Riko should keep mapping-like items and iterable streams as the core contract. |
| [RssPercolator](https://github.com/olviko/RssPercolator) | Asynchronous multi-source fetching, multiple destinations, and filters evaluated in declared order. | Feed-specific concepts embedded in the core engine. RSS and Atom should remain modules or recipe packages. |
| [Plagger](https://github.com/miyagawa/plagger) | A plugin-oriented architecture, declarative YAML recipes, and a gallery of end-to-end examples such as RSS-to-email. | An implicit global hook system with weakly defined contracts between plugins. Riko modules should have explicit inputs, outputs, configuration, and capabilities. |
| [Turtle](https://github.com/iopipe/turtle) | A boundary between composition and execution drivers, local and remote function chaining, scatter/gather, and attention to package integrity. | Cloud-first architecture, automatic execution of downloaded code, or provider-specific behavior in the core package. |
| [node-machine](https://github.com/node-machine/machine) | Machine-readable function definitions, typed inputs, named exits, generated documentation and tests, static analysis, and reusable packs. | Multiple overlapping invocation styles and deferred-control APIs that make execution semantics harder to reason about. |

### Primary decision

The highest-value idea is a self-describing module contract. It unlocks plugin discovery, workflow validation, generated documentation, CLI inspection, GUI forms, compatibility checks, and remote execution boundaries. It must be implemented before a visual editor or distributed driver.

### Secondary decisions

- Use Python package metadata entry points for third-party module discovery rather than scanning arbitrary import paths.
- Use a versioned JSON-compatible workflow schema as the canonical storage format. YAML may be accepted as an authoring representation, but it must normalize to the same model.
- Keep visual tooling in a separate repository and make it consume exported schemas and workflow files.
- Add observability through optional hooks and APIs. The core library must not require an observability backend.
- Treat remote execution as an experimental adapter after local lifecycle, serialization, and error semantics are stable.

## Current baseline

The `features` branch already provides much of the runtime foundation:

- Python 3.12 or newer, with sync and async APIs.
- Bounded async concurrency and backpressure-oriented execution.
- A runtime-derived built-in module catalog.
- Public `riko` and `riko.api` surfaces plus a supported `riko.ext` authoring surface.
- Processor, operator, and splitter decorators.
- JSON pipe definitions, a compact DAG format, in-process execution, and Python code generation.
- Split, send, and receive primitives for fan-out and pub/sub-style routing.
- CLI commands for execution, compilation, DAG conversion, configuration generation, and benchmarking.

Known gaps that directly affect this roadmap:

- `riko.ext.register` is intentionally deferred; there is no public runtime plugin registry yet.
- Module metadata does not yet provide a complete machine-readable input/output/configuration contract.
- The compact DAG format cannot represent all named ports and fan-in cases.
- Workflow schema versioning and migrations are not defined.
- Execution telemetry, retry policies, checkpointing, and remote drivers are not stable public capabilities.
- Split routing and dynamic-key correctness issues remain open.

## Roadmap principles

1. **Contracts before interfaces.** CLI, GUI, plugins, and drivers must consume one contract rather than invent parallel models.
2. **Local semantics are authoritative.** Generated code and every execution driver must match the in-process executor.
3. **Explicit ports, errors, and side effects.** A module contract must state what it accepts, emits, mutates, and depends on.
4. **Optional integrations stay optional.** Plugin loading, OpenTelemetry, cloud SDKs, and protocol clients must not inflate the minimal install.
5. **Version data, not implementation details.** Workflow files and plugin contracts need schema versions and migrations; private Python objects are not a durable storage format.
6. **Secure by default.** Installing a Python package is a trust decision. Riko must not silently fetch or execute remote code.
7. **Evidence before optimization.** Performance work must be tied to reproducible workloads and measured regressions.

## Phase 0: Correctness and release baseline

Goal: establish a trustworthy local runtime before expanding the public surface.

### Deliverables

- Resolve [#61](https://github.com/nerevu/riko/issues/61): make split routing port-aware and identical between `build_pipeline` and generated code.
- Resolve [#60](https://github.com/nerevu/riko/issues/60): preserve literal data-derived keys without breaking configured dotted paths.
- Expand [#11](https://github.com/nerevu/riko/issues/11) into repeatable benchmark scenarios for throughput, peak memory, file descriptors, open sockets, and async concurrency.
- Define executor-versus-codegen parity tests for linear, fan-in, split, nested split, send/receive, sync, and async workflows.
- Publish a compatibility matrix for the stable API, extension API, workflow format, and generated Python.
- Align CI branch targets and release metadata with the actual release branch policy.

### Acceptance criteria

- Every checked-in workflow fixture produces equivalent results in the in-process and generated-code paths.
- Invalid port references and unsupported graphs fail before execution with a public Riko exception.
- Benchmark commands emit machine-readable results suitable for CI comparison.
- No new extension, workflow-schema, or driver API is marked stable while correctness blockers remain open.

## Phase 1: Module contract v1

Goal: make every built-in and third-party module self-describing.

This phase borrows named ports from Pypes and machine-readable definitions from node-machine.

### Contract fields

Each module definition should expose at least:

- stable name and contract version;
- module type and supported subtypes;
- sync and async availability;
- input ports and output ports, including cardinality;
- configuration schema, defaults, required values, and deprecations;
- item field expectations and emitted field behavior where knowable;
- named error outcomes or documented exception classes;
- side-effect flags such as network, filesystem, subprocess, prompt, and nondeterministic;
- streaming characteristics such as lazy, buffering, materializing, unbounded, and order-preserving;
- concurrency safety and cancellation support;
- short description, examples, and documentation URL;
- distribution name and version for external modules.

### Deliverables

- Extend `ModuleMetadata` or introduce a versioned `ModuleDefinition` type.
- Generate configuration JSON Schema from Riko's typed config definitions.
- Add `riko module list`, `riko module show NAME`, and `riko module schema NAME` commands.
- Add `riko pipeline validate FILE` and `riko pipeline explain FILE` commands.
- Export the built-in catalog as deterministic JSON for documentation and GUI consumers.
- Add a contract conformance test helper for module authors.

### Acceptance criteria

- All built-in modules pass the same contract conformance suite.
- The module catalog can be generated without importing optional dependencies that are not installed.
- Documentation tables are generated from the catalog rather than maintained manually.
- Unknown configuration keys, missing required values, invalid port connections, and incompatible capabilities produce actionable validation errors.

## Phase 2: Plugin ecosystem v1

Goal: complete [#9](https://github.com/nerevu/riko/issues/9) with a small, explicit, package-native plugin system.

This phase borrows the plugin ecosystem idea from Plagger and the reusable pack concept from node-machine.

### Discovery model

Use Python package metadata entry points. A plugin distribution should declare modules under a dedicated group such as:

```toml
[project.entry-points."riko.modules"]
example = "riko_example:modules"
```

The loaded object should return module definitions or invoke a public registration function. Import-path scanning and arbitrary code downloads should not be the default discovery mechanism.

### Deliverables

- Add `riko.ext.register` and a documented registry protocol.
- Discover installed plugins through `importlib.metadata.entry_points()`.
- Define deterministic name-conflict rules and explicit override behavior.
- Record provider distribution, provider version, contract version, and load errors.
- Add `riko plugin list`, `riko plugin inspect`, and `riko plugin doctor` commands.
- Publish a minimal plugin template with tests, typing, packaging, and documentation.
- Support disabling plugins by name or distribution for troubleshooting.
- Add compatibility checks against Riko and module-contract versions.

### Security requirements

- Never install plugins automatically while loading a workflow.
- Never download or execute remote code from a workflow definition.
- Display plugin provenance in validation, execution plans, and errors.
- Permit applications to supply an allowlist or denylist.
- Document that an installed Python plugin has the same process privileges as the host application.

### Acceptance criteria

- A separately distributed package can add a processor, operator, or splitter without modifying `riko.modules`.
- Plugin failures are isolated during discovery and reported without hiding healthy modules.
- A workflow can be validated against an installed plugin without executing the workflow.
- Built-in and third-party modules use the same definition and conformance APIs.

## Phase 3: Workflow specification v1

Goal: resolve the storage and interoperability concerns in [#16](https://github.com/nerevu/riko/issues/16) and provide the stable foundation required by [#17](https://github.com/nerevu/riko/issues/17).

This phase combines Riko's existing JSON and compact DAG formats with the declarative recipe approach demonstrated by Plagger and RssPercolator.

### Design requirements

- Canonical, JSON-compatible data model.
- Explicit schema version.
- Stable node identifiers and explicit source and target ports.
- Complete support for linear flows, fan-out, fan-in, split outputs, named inputs, and terminal outputs.
- Optional human-friendly YAML parsing that normalizes to the canonical model.
- Environment-independent core document; secrets and runtime inputs are references, not embedded credentials.
- Deterministic serialization for reviewable diffs.
- Forward migration tools and clear rejection of unsupported future versions.

### Deliverables

- Publish a JSON Schema 2020-12 document for workflow v1.
- Define a canonical normalized workflow model and retain the compact DAG as a convenience syntax.
- Add `format_version`, optional metadata, parameters, resource references, and declared outputs.
- Add round-trip tests for JSON, YAML, normalized objects, and generated Python.
- Add `riko pipeline format`, `migrate`, `diff`, and `graph` commands.
- Define a recipe directory convention with example input data and expected output assertions.
- Add workflow signing or digest fields only as integrity metadata; do not imply trust in referenced code.

### Acceptance criteria

- Every representable in-process pipeline can round-trip through workflow v1 without losing port information.
- Workflow files remain usable across patch and minor Riko releases through validation or documented migration.
- A GUI can render node forms and connections using only the workflow schema and module catalog.
- Secrets are supplied through runtime inputs or application-specific resolvers, not serialized into examples.

## Phase 4: Runtime policies and observability

Goal: make pipeline behavior diagnosable and controllable without coupling Riko to a scheduler or hosted service.

This phase borrows explicit lifecycle ideas from Mario and operational granularity from Turtle.

### Deliverables

- Define lifecycle events for pipeline, module, item batch, retry, cancellation, and resource closure.
- Add optional callbacks or an event sink to `Context`.
- Add structured execution records with workflow ID, run ID, module ID, provider, duration, counts, and outcome.
- Provide optional OpenTelemetry API integration for traces and metrics without requiring an SDK in the core install.
- Add per-module timeout, retry, backoff, and failure-policy configuration where semantics are safe.
- Define cancellation propagation and resource cleanup requirements for sync and async modules.
- Add bounded diagnostic sampling so item payloads are not logged by default.
- Add execution-plan estimates for known buffering or materialization boundaries.

### Required metrics

- items received, emitted, skipped, and failed;
- wall time and active processing time;
- queue depth or in-flight task count for bounded async execution;
- retry and timeout counts;
- peak memory for benchmark workloads;
- bytes read and written when the adapter can report them;
- resource closure failures.

### Acceptance criteria

- Instrumentation disabled has negligible overhead in benchmarked common paths.
- Instrumentation never records full item payloads unless the application explicitly opts in.
- Cancellation closes owned files, responses, task groups, and subprocess handles.
- Retry policies cannot silently repeat modules declared non-idempotent without explicit user approval.

## Phase 5: Adapter and connector packages

Goal: broaden useful inputs and outputs without turning the core package into a protocol collection.

This phase borrows Mario's file-like and subprocess adapters, RssPercolator's multi-source/multi-destination model, and Plagger's end-to-end recipes.

### Deliverables

- Define source and sink adapter protocols around the module contract.
- Add first-class local adapters for file-like objects, iterables, async iterables, standard input/output, and subprocess streams.
- Standardize multi-source merge and multi-sink broadcast behavior, including ordering and failure policy.
- Move specialized protocols from [#19](https://github.com/nerevu/riko/issues/19) into separately versioned plugin packages unless they are essential to the core distribution.
- Publish recipe packs for feed aggregation, file conversion, HTTP enrichment, and notification workflows.
- Add connector contract tests using local fixtures and fake servers rather than live external services.

### Prioritization rule

A connector belongs in the core package only when it is broadly useful, has a small dependency footprint, and can be tested deterministically. Otherwise it belongs in a plugin distribution.

### Acceptance criteria

- Adapters obey the same cancellation, closure, provenance, and observability contracts as built-in modules.
- Multi-source and multi-sink workflows have documented ordering and partial-failure behavior.
- Optional protocol packages can release independently from Riko core.

## Phase 6: Experimental execution drivers

Goal: prove that a workflow can execute outside the local process without changing its logical result.

This phase borrows Turtle's driver boundary and scatter/gather model, but deliberately postpones cloud-specific implementation until local contracts are stable.

### Driver contract

A driver should accept a validated execution plan and report standardized outcomes. It must not reinterpret graph structure or module semantics.

### Deliverables

- Define an experimental driver interface outside the stable API tier.
- Implement a reference local driver that delegates to the existing executor.
- Add a process-pool driver for modules explicitly marked serializable and process-safe.
- Define serialization boundaries for configuration, items, errors, and plugin provenance.
- Add scatter/gather planning for eligible stateless modules with bounded concurrency.
- Specify idempotency, retries, cancellation, artifact transfer, and result ordering.
- Evaluate one remote proof of concept only after the process driver passes conformance tests.

### Non-goals for the first driver release

- A scheduler service.
- A workflow database.
- Automatic cloud deployment.
- Transparent execution of arbitrary installed modules on untrusted workers.
- Exactly-once processing claims.

### Acceptance criteria

- Local and process drivers pass a shared semantic conformance suite.
- Unsupported modules are rejected during planning rather than failing after partial execution.
- Driver selection does not change workflow files.
- Remote proof-of-concept code remains optional and experimental until security and compatibility policies are documented.

## Phase 7: Visual tooling

Goal: address [#17](https://github.com/nerevu/riko/issues/17) without coupling UI concerns to the engine.

Pypes demonstrates the value of a graph editor, while the historical feedback around Yahoo! Pipes shows that large visual pipelines can become harder to manage than text. The GUI should therefore complement, not replace, reviewable workflow files.

### Deliverables

- Create a separate repository for the editor.
- Generate the node palette, forms, validation messages, and help text from module definitions.
- Import and export workflow v1 without private Riko objects.
- Provide graph validation before execution.
- Provide a read-only execution plan and event stream before adding editing-time execution.
- Support source view, normalized diff view, search, grouping, and subflow collapse for large pipelines.
- Keep secrets in the host application or runtime input system.

### Acceptance criteria

- The editor can work against a static exported catalog without importing Riko in the browser.
- A file edited in the GUI produces a deterministic, reviewable workflow diff.
- All validation rules also exist in the Python library and CLI.
- The editor is optional and versioned independently from Riko core.

## Phase 8: Ecosystem and 1.0 readiness

Goal: make the contracts dependable enough for independent module and tooling authors.

### Deliverables

- Publish the stable API, extension API, module-contract, workflow-schema, and compatibility policies.
- Publish a plugin author guide, workflow author guide, and driver author guide.
- Add a conformance badge process based on automated tests, not manual approval.
- Maintain a curated recipe gallery with reproducible fixtures.
- Publish benchmark history and regression thresholds.
- Define deprecation and migration windows.
- Release 1.0 only after at least one external plugin and one external workflow consumer validate the contracts.

### 1.0 exit criteria

- No open P0 correctness issue in execution, routing, lifecycle, or serialization.
- Stable workflow v1 with migrations and schema validation.
- Stable module contract and plugin discovery API.
- Documented security and provenance model.
- Reproducible benchmarks and supported Python-version CI.
- Generated code and the local executor pass the same conformance corpus.

## Proposed issue decomposition

The following issues can turn this roadmap into independently reviewable work:

1. Define `ModuleDefinition` v1 and compatibility rules.
2. Export module configuration JSON Schema.
3. Add module inspection and pipeline validation CLI commands.
4. Implement entry-point plugin discovery.
5. Add `riko.ext.register` and conflict handling.
6. Publish the third-party plugin template and conformance suite.
7. Define workflow v1 JSON Schema and normalization rules.
8. Add workflow migration and deterministic formatting.
9. Add lifecycle events and structured execution records.
10. Add optional OpenTelemetry API hooks.
11. Define retry, timeout, cancellation, and idempotency policies.
12. Add source/sink adapter protocols and subprocess adapters.
13. Define the experimental execution-driver protocol.
14. Implement local and process driver conformance tests.
15. Create the separate visual-editor repository after phases 1 through 3 are complete.

## Dependency map

```text
Phase 0: correctness and parity
    |
    v
Phase 1: module contract
    |--------------------|
    v                    v
Phase 2: plugins      Phase 3: workflow spec
    |                    |
    |----------|---------|
               v
Phase 4: runtime policies and observability
               |
               v
Phase 5: adapters and connector packages
               |
               v
Phase 6: execution drivers
               |
               v
Phase 7: visual tooling
               |
               v
Phase 8: ecosystem and 1.0
```

The GUI may be prototyped earlier against exported static schemas, but it should not define or fork the runtime contracts.

## Explicitly deferred

The following are not recommended until the preceding contracts are stable:

- a hosted Riko control plane;
- automatic package installation from workflow files;
- a marketplace that executes unreviewed modules;
- exactly-once delivery guarantees;
- durable distributed checkpointing;
- transparent cross-language modules;
- a built-in drag-and-drop web server;
- cloud-provider SDK dependencies in the core install.

These may become separate projects after there is demonstrated demand and a secure execution model.

## Research references

### Issue sources

- [Issue #10: similar projects](https://github.com/nerevu/riko/issues/10)
- [Issue #9: plugin capability](https://github.com/nerevu/riko/issues/9)
- [Issue #11: benchmarks and profiles](https://github.com/nerevu/riko/issues/11)
- [Issue #16: pipeline file format](https://github.com/nerevu/riko/issues/16)
- [Issue #17: GUI strategy](https://github.com/nerevu/riko/issues/17)
- [Issue #19: additional protocols](https://github.com/nerevu/riko/issues/19)
- [Issue #60: literal data-derived keys](https://github.com/nerevu/riko/issues/60)
- [Issue #61: split routing parity](https://github.com/nerevu/riko/issues/61)

### Project sources

- [Pypes repository](https://github.com/fullscale/pypes)
- [Pypes core description](https://github.com/fullscale/pypes/blob/master/core/README.txt)
- [Mario README](https://github.com/colinmarc/mario/blob/master/README.md)
- [RssPercolator README](https://github.com/olviko/RssPercolator/blob/master/README.md)
- [Plagger RSS-to-email recipe](https://github.com/miyagawa/plagger/blob/master/examples/rss2email.yaml)
- [Turtle README](https://github.com/iopipe/turtle/blob/master/README.md)
- [node-machine README](https://github.com/node-machine/machine/blob/master/README.md)

### Standards and implementation references

- [Python plugin discovery guidance](https://packaging.python.org/en/latest/guides/creating-and-discovering-plugins/)
- [Python entry points specification](https://packaging.python.org/en/latest/specifications/entry-points/)
- [JSON Schema specification](https://json-schema.org/specification)
- [OpenTelemetry Python instrumentation](https://opentelemetry.io/docs/languages/python/instrumentation/)

**Comment by reubano** [comment]
# Roadmap Addendum: Competitive Positioning and Borrowing Priorities

Status: Proposed addendum  
Research date: 2026-08-07  
Related roadmap: existing 2026 roadmap comment in this issue

This addendum updates the roadmap with lessons from a closer comparison against modern Python ETL and stream-processing libraries: [Bonobo](https://www.bonobo-project.org/), [petl](https://petl.readthedocs.io/), [Singer](https://www.singer.io/), [Streamz](https://streamz.readthedocs.io/), [Bytewax](https://docs.bytewax.io/), and [FastPipe](https://pypi.org/project/fastpipe/).

## Updated product positioning

Riko should **not** position itself primarily around running locally, in-process, or without a scheduler/cluster. Bonobo, petl, Streamz, FastPipe, and even single-worker Bytewax can all run directly from ordinary Python.

The more defensible differentiator is the combination of:

- a broad, configuration-driven vocabulary of reusable pipes;
- first-class RSS/Atom and web-content processing;
- dictionary-like record streams with implicit item processing;
- synchronous and native async APIs over the same pipe vocabulary;
- local thread/process execution for suitable item-level work;
- lazy fan-out through `send` / `receive` pub/sub;
- pipelines that can be represented as JSON-compatible workflow definitions;
- workflow inspection, execution, and compilation into Python.

A useful shorthand remains that Riko sits between **“just write a Python loop”** and **“adopt a dedicated data-processing platform”**, but locality alone is not the differentiator.

## What the newer comparisons reveal

| Project | High-value capability to learn from | Why it matters to Riko | What Riko should not become |
| --- | --- | --- | --- |
| [Bonobo](https://www.bonobo-project.org/) | ETL DAGs with injectable runtime services/resources | Strengthens the case for explicit resource/context contracts and easier test/runtime substitution | A framework whose primary abstraction is dependency injection or graph execution |
| [petl](https://petl.readthedocs.io/) | Rich, purpose-built table algebra and data-quality primitives | Suggests expanding high-value built-ins where they remove repetitive user code, especially reconciliation/validation-style operations | A dataframe/table engine or relational query system |
| [Singer](https://www.singer.io/) | Standardized tap/target contracts, schema messages, and incremental replication state | Suggests a clean connector boundary and reusable source/sink contracts; useful for import/export adapters | A connector-replication protocol whose main job is source-to-destination movement |
| [Streamz](https://streamz.readthedocs.io/) | Continuous push-oriented graphs, live windows, branching, and backpressure | Provides useful ideas for bounded streaming semantics, temporal operators, and explicit stream behavior metadata | A reactive/event framework where push semantics replace Riko's iterator-oriented core |
| [Bytewax](https://docs.bytewax.io/) | Durable keyed state, recovery, partitioned sources, and distributed workers | Clarifies the boundary between lightweight local state and a true streaming runtime; informs adapter interfaces if remote execution is ever added | A distributed stateful engine with checkpoint/recovery machinery in core |
| [FastPipe](https://pypi.org/project/fastpipe/) | Explicit per-stage thread/process/async execution modes with adapters between stages | Suggests eventually separating execution policy from pipeline-wide mode and making stage capabilities machine-readable | A low-level concurrency framework with a deliberately tiny transformation vocabulary |
| Riko | Configuration-driven record-stream vocabulary + feeds/web + sync/async/local parallelism + workflows-as-data | Preserve and deepen this intersection | Chasing every adjacent project's strongest feature until the core loses coherence |

## Roadmap implications

### 1. Make module capabilities machine-readable

The existing **Module contract v1** remains the highest-leverage roadmap item, but the competitive review suggests several capabilities should be explicit rather than inferred:

- supported execution modes: sync, async, thread-safe, process-safe;
- source / transform / aggregate / sink role;
- lazy vs buffering vs materializing behavior;
- bounded/unbounded input suitability;
- ordering guarantees;
- stateful/stateless behavior;
- side effects and resource requirements;
- fan-in/fan-out cardinality;
- schema or field expectations where knowable;
- checkpoint/recovery capability, even if the built-in value is `none`;
- temporal/windowing capability, if introduced later.

This metadata would let the CLI, compiler, validator, GUI, and future execution adapters reason about a workflow without hard-coding module names.

### 2. Treat connector contracts as a separate concern from transformation pipes

Singer reinforces that **source/sink interoperability is its own useful abstraction**.

Riko should consider a small connector contract for reusable external-system boundaries while keeping transformations as ordinary pipes. A connector layer could eventually support:

- standardized discovery and configuration;
- schema/field metadata where available;
- incremental cursor/state handoff for sources that support it;
- explicit resource ownership and close semantics;
- source/sink provenance in workflow inspection;
- optional adapters to Singer taps/targets rather than reimplementing their ecosystem.

This belongs naturally with the plugin and adapter phases, not in the core record model.

### 3. Prefer high-value built-in operations over generic API breadth

petl demonstrates the value of domain-specific primitives that collapse otherwise repetitive application code.

Riko should continue adding built-ins selectively when an operation:

1. is common across multiple workflows;
2. has non-trivial streaming/materialization semantics worth standardizing;
3. benefits from configuration/metadata/compilation support; and
4. can be implemented without turning Riko into a dataframe engine.

Possible areas to evaluate include reconciliation/conflict detection, validation, richer grouping/aggregation, and schema-oriented record operations.

The bar should be higher than “this can be another wrapper around a one-line Python function.” `udf` already covers one-off application logic.

### 4. Keep temporal/stateful streaming deliberately bounded in scope

Streamz and Bytewax show two very different levels of streaming sophistication:

- Streamz: continuous graphs, backpressure, live windows, reactive composition;
- Bytewax: keyed state, recovery, partitioning, worker scaling, distributed execution.

Riko can borrow **semantics and metadata** without inheriting the full runtime complexity.

Near-term candidates:

- document and expose backpressure/buffering boundaries consistently;
- make bounded concurrency settings inspectable in execution plans;
- evaluate finite sliding/count windows as ordinary operators;
- define what “stateful” means for a Riko module;
- keep durable checkpointing/recovery out of core unless a future driver contract proves it can remain optional.

### 5. Explore per-stage execution policy only after module contracts are stable

FastPipe exposes a meaningful capability Riko currently does not: one pipeline can deliberately move through async, process, and thread stages with adapters between them.

Riko should not copy this immediately. First, the module contract needs enough information to answer whether a stage is safe and useful under a given execution mode.

After that foundation exists, investigate a workflow-level execution policy such as:

```text
fetch       -> async / high concurrency
parse       -> process pool
store       -> thread pool
```

Any implementation should preserve Riko's existing simple defaults. Per-stage execution should be an explicit optimization, not required boilerplate.

### 6. Strengthen `Context` as the resource boundary

Bonobo's service injection is useful because transformations can remain independent of environment-specific resources.

Riko already has `Context`; rather than inventing a parallel service container, continue evolving `Context` into the explicit runtime boundary for:

- runtime inputs;
- execution mode/policy;
- optional resource resolvers;
- instrumentation/event sinks;
- connector state;
- cancellation/deadline information;
- future driver-specific resources.

The module contract should declare what it expects from context rather than reaching into global state.

### 7. Keep workflows-as-data a first-class differentiator

The closer comparison reinforces that Riko's JSON-compatible workflow definitions are strategically valuable.

The roadmap should continue prioritizing:

- a versioned canonical workflow schema;
- deterministic serialization;
- complete port representation;
- dependency/capability inspection without execution;
- round-trip parity between workflow data, in-process execution, and generated Python;
- tooling that can consume the same workflow/module contracts without importing private implementation details.

This creates a clean foundation for CLIs, GUIs, generated workflows, policy checking, and integration with larger orchestration systems without making any of those systems part of Riko core.

## Explicit non-goals clarified by the comparison

Riko should not attempt to match every adjacent project's strongest capability.

Unless the product direction changes materially, the following should remain outside core:

- Bytewax-style distributed worker coordination and durable recovery;
- Streamz-style full reactive dataframe semantics;
- petl-style comprehensive relational/table algebra;
- Singer's full connector catalog and replication protocol ecosystem;
- Bonobo-style service-container-centric application architecture;
- FastPipe-style concurrency primitives as the primary user-facing abstraction;
- task scheduling/orchestration, durable retries, hosted workflow management, or cluster deployment.

Where useful, Riko should interoperate with these systems rather than absorb them.

## Proposed prioritization changes

The existing phase order still largely holds. This addendum suggests the following refinements:

1. **Module contract v1:** add execution, state, streaming, resource, and connector capability metadata.
2. **Plugin ecosystem v1:** support connector/source/sink plugins under the same provenance and compatibility rules as transformation modules.
3. **Workflow specification v1:** ensure execution policy and connector state references can be represented without coupling the schema to a particular runtime.
4. **Runtime policies:** make buffering, backpressure, retries, cancellation, and execution choices inspectable.
5. **Adapter/connector packages:** evaluate Singer interoperability before inventing a large native connector ecosystem.
6. **Execution drivers:** only after contracts are stable, evaluate per-stage execution policy and optional remote/distributed adapters.
7. **Built-in vocabulary:** continuously evaluate high-value operations that reduce boilerplate while preserving Riko's record-stream model.

## Updated product test

For future feature proposals, ask:

> Does this make Riko better at expressing, inspecting, or executing configuration-driven record-stream transformations, or does it mainly reproduce infrastructure another project already specializes in?

If the answer is the latter, prefer an adapter or integration over expanding the core.

The goal is not to make Riko a smaller version of every ETL/streaming system. The goal is to make its particular intersection — **reusable configured pipes, rich feed/web and record transformations, multiple local execution models, lazy fan-out, and workflows-as-data** — increasingly coherent and useful.
